File: android/biangbiang-ui/src/main/java/dev/veeso/biangbiangui/services/TextProcessingEngine.kt
package dev.veeso.biangbiangui.services

import dev.veeso.biangbiangui.protocols.Transliterator

/**
 * Detects script spans via a regex built from `scriptRanges`, transliterates
 * each span using an injected [Transliterator], and applies the reference
 * spacing and cleanup rules.
 *
 * Ported from the reference BiangBiang Hanzi Android `TextProcessor`. The
 * generalisations mirror iOS Phase 2 `TextProcessingEngine`:
 * - The hard-coded CJK regex is replaced by a character class built from the
 *   supplied [scriptRanges] (`TextProcessor` -> `TextProcessingEngine`).
 * - The `pinyin4j`/`JyutpingDictionary` mode switch is replaced by a single
 *   injected [Transliterator] (one span in, Latin out).
 * - The post-pass cleanup (strip whitespace before punctuation, collapse
 *   double spaces) matches the canonical iOS surface.
 */
class TextProcessingEngine(
    scriptRanges: List<UIntRange>,
    private val transliterator: Transliterator,
) {
    // Build a character-class pattern: [\uXXXX-\uYYYY...]+
    // All CJK ranges in use are BMP, so 4-hex \u escapes are unambiguous;
    // this matches the iOS NSRegularExpression class-body construction.
    private val pattern: Regex = run {
        val classBody = scriptRanges.joinToString("") { range ->
            "\\u%04X-\\u%04X".format(range.first.toInt(), range.last.toInt())
        }
        Regex("[$classBody]+")
    }

    /**
     * Transliterates all script spans in [text], preserving non-script
     * characters and applying the reference leading/trailing spacing rules and
     * post-pass cleanup.
     *
     * @return the processed string, or `null` if [text] contained no script
     *   spans.
     */
    fun process(text: String): String? {
        if (!containsScript(text)) return null

        // Iterate matches in reverse so replacing a span does not invalidate
        // the offsets of earlier (lower-index) spans.
        val matches = pattern.findAll(text).toList().asReversed()
        val result = StringBuilder(text)

        for (match in matches) {
            val start = match.range.first
            val end = match.range.last + 1

            val span = result.substring(start, end)
            val transliterated = transliterator.transliterate(span)

            // Leading space: insert unless the span is at string start or the
            // immediately preceding character is a space or sentence punctuation.
            val needsLeadingSpace = when {
                start == 0 -> false
                result[start - 1] == ' ' -> false
                ".,!?;:".contains(result[start - 1]) -> false
                else -> true
            }

            // Trailing space: insert only when the immediately following
            // character is an ASCII letter or digit.
            val needsTrailingSpace = when {
                end >= result.length -> false
                else -> {
                    val next = result[end]
                    next.code < 128 && (next.isLetter() || next.isDigit())
                }
            }

            val replacement = buildString {
                if (needsLeadingSpace) append(' ')
                append(transliterated)
                if (needsTrailingSpace) append(' ')
            }

            result.replace(start, end, replacement)
        }

        // Post-pass 1: remove whitespace immediately before punctuation.
        // Post-pass 2: collapse runs of 2+ whitespace into a single space.
        return result
            .toString()
            .replace(Regex("\\s+([.,!?;:])"), "$1")
            .replace(Regex("\\s{2,}"), " ")
            .trim()
    }

    /** Returns `true` iff [text] contains at least one matchable script span. */
    fun containsScript(text: String): Boolean = pattern.containsMatchIn(text)
}

File: Sources/BiangBiangUI/Services/TextProcessingEngine.swift
import Foundation

/// Detects script spans via a regex built from `scriptRanges`, transliterates each span
/// using an injected `Transliterator`, and applies the reference spacing and cleanup rules.
///
/// The regex is immutable after initialisation; `@unchecked Sendable` is used only because
/// `NSRegularExpression` does not conform to `Sendable`, but is thread-safe when read-only.
public final class TextProcessingEngine: @unchecked Sendable {
    // MARK: - Stored properties

    private let regex: NSRegularExpression
    private let transliterator: any Transliterator

    // MARK: - Init

    /// Builds a `TextProcessingEngine` whose span detector matches any run of characters
    /// whose Unicode scalar values fall within any of `scriptRanges`.
    ///
    /// - Parameters:
    ///   - scriptRanges: One or more closed Unicode scalar ranges (e.g. `[0x4E00...0x9FFF]`).
    ///   - transliterator: Strategy that converts an isolated script span to a Latin string.
    public init(scriptRanges: [ClosedRange<UInt32>], transliterator: any Transliterator) {
        // Build a character-class pattern: [\uXXXX-\uYYYYꪪ-뮻...]+
        // NSRegularExpression uses ICU regex; \uXXXX (4-hex) escapes work for BMP code points,
        // and \U{XXXXXX} (with braces) for supplementary planes — but all CJK ranges used here
        // are BMP so 4-hex is sufficient and unambiguous.
        let classBody = scriptRanges.map { range in
            String(format: "\\u%04X-\\u%04X", range.lowerBound, range.upperBound)
        }.joined()

        let pattern = "[\(classBody)]+"
        // swiftlint:disable:next force_try — pattern is machine-generated from UInt32 values
        regex = try! NSRegularExpression(pattern: pattern, options: [])
        self.transliterator = transliterator
    }

    // MARK: - Public API

    /// Transliterates all script spans in `text`, preserving non-script characters and
    /// applying the reference leading/trailing spacing rules and post-pass cleanup.
    ///
    /// - Returns: The processed string, or `nil` if `text` contained no script spans.
    public func process(_ text: String) -> String? {
        guard containsScript(text) else { return nil }

        let nsRange = NSRange(text.startIndex..., in: text)
        var result = text

        // Iterate matches in reverse so that replacing a span does not invalidate
        // the character offsets of earlier (lower-index) spans.
        let matches = regex.matches(in: text, range: nsRange).reversed()

        for match in matches {
            guard let range = Range(match.range, in: result) else { continue }

            let span = String(result[range])
            let transliterated = transliterator.transliterate(span)

            // Leading space: insert unless the span is at the very start of the string
            // or the immediately preceding character is a space or sentence punctuation.
            let needsLeadingSpace: Bool = {
                guard range.lowerBound != result.startIndex else { return false }
                let prevChar = result[result.index(before: range.lowerBound)]
                return prevChar != " " && !".,!?;:".contains(prevChar)
            }()

            // Trailing space: insert only when the immediately following character is
            // an ASCII letter or digit (prevents gluing the romanisation to the next word).
            let needsTrailingSpace: Bool = {
                guard range.upperBound < result.endIndex else { return false }
                let nextChar = result[range.upperBound]
                return nextChar.isASCII && (nextChar.isLetter || nextChar.isNumber)
            }()

            let replacement =
                (needsLeadingSpace ? " " : "") +
                transliterated +
                (needsTrailingSpace ? " " : "")

            result.replaceSubrange(range, with: replacement)
        }

        // Post-pass 1: remove any whitespace that ended up immediately before punctuation.
        result = result.replacingOccurrences(
            of: "\\s+([.,!?;:])",
            with: "$1",
            options: .regularExpression
        )

        // Post-pass 2: collapse runs of two or more spaces into a single space.
        result = result.replacingOccurrences(
            of: "\\s{2,}",
            with: " ",
            options: .regularExpression
        )

        return result.trimmingCharacters(in: .whitespaces)
    }

    /// Returns `true` iff `text` contains at least one script span that the regex can match.
    public func containsScript(_ text: String) -> Bool {
        let nsRange = NSRange(text.startIndex..., in: text)
        return regex.firstMatch(in: text, range: nsRange) != nil
    }
}
